✨ Add `.load()` for artifacts holding `SpatialData` objects

Zethson opened the pull request:
Fixes #2513 

- Adds support for loading `SpatialData` `Artifacts` into memory
- Adds a new `ScverseDataStructures` type
- Adds a new `_compat.py` module which has utility functions to execute code from functions that require modules to be installed and available
- Introduces a new `data_fixtures` module that will eventually consolidate data fixtures across tests as much as reasonable
- Adds more tests for various `.load()` functions

sunnyosun commented on lamindb/core/loaders.py:
```diff
@@ -172,7 +175,7 @@ def load_rds(path: UPathStr) -> UPathStr:
 
 
-def load_to_memory(filepath: UPathStr, **kwargs):
+def load_to_memory(filepath: UPathStr, **kwargs) -> None | UPathStr:
```
Any?

sunnyosun replied:
It never returns None, does it?

sunnyosun commented:
Is it possible to get more lines covered?
<img width="462" alt="Screenshot 2025-03-05 at 21 26 01" src="https://github.com/user-attachments/assets/fef01750-ac59-4b6b-9702-a10bbc77afda" />

sunnyosun commented:
I'm not so sure about `ScverseDataStructures` as I'd prefer to see them listed.

Zethson commented:
> Is it possible to get more lines covered? <img alt="Screenshot 2025-03-05 at 21 26 01" width="462" src="https://private-user-images.githubusercontent.com/38218185/419637322-fef01750-ac59-4b6b-9702-a10bbc77afda.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDEyMDY3OTksIm5iZiI6MTc0MTIwNjQ5OSwicGF0aCI6Ii8zODIxODE4NS80MTk2MzczMjItZmVmMDE3NTAtYWM1OS00YjZiLTk3MDItYTEwYmJjNzdhZmRhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMDUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzA1VDIwMjgxOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWU5YjZlNDRkZDMxNmQyNmRhM2RhNWZjOGYyODkzNDQzYzQ2NWViNmRiYTA5Y2FmMTdlNWY5NzY2MmY0N2UxNzMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.WmZBhcrts_UKQY2tQaNwLVRwidj-kVg0uubu1ff6eJU">

Two of those files are in coverage ignore but it apparently doesn't work. Sigh.

sunnyosun commented on lamindb/core/loaders.py:
```diff
@@ -156,7 +158,8 @@ def load_rds(path: UPathStr) -> UPathStr:
     ".fcs": load_fcs,
-    ".zarr": load_anndata_zarr,
+    ".zarr": load_zarr,
+    ".spatialdata.zarr": load_zarr,
```
Why do we need `.spatialdata.zarr` here since `.zarr` uses the same reader?

Zethson replied:
Hmm, using `.spatialdata.zarr` is more explicit but sure - in this case we can omit it :) Done.

sunnyosun reviewed (approved):
Good from my side! But I'd suggest either @Koncopd or @falexwolf to also take a look!

Koncopd commented:
I will take a look tmr.

Zethson commented on lamindb/core/_compat.py:
```diff
@@ -0,0 +1,60 @@
+    return spec is not None
+
+
+def with_package(package_name: str, operation: Callable[[Any], T]) -> T:
```
Since at least one more person will look at this PR, I make a note here. I know that this is a bit ugly but allows us to write some safer code. Else, we also always have to write lots of statements where we check for available packages, the type of whatever we're looking for, and a PLEASE INSTALL XYZ message. It gets old quick.

Happy to revert if ugly or we can solve it differently. Else, I'll use it also in other places.

falexwolf commented:
> Adds support for loading SpatialData Artifacts into memory

Curious: Is that a typical usage pattern? Won't most people use `.open()`? (Supporting it is good either way even if it's only going to be used for small objects.)

> Adds a new ScverseDataStructures type

I like it.

falexwolf commented on lamindb/core/types.py:
```diff
@@ -7,3 +11,23 @@
+    from anndata import AnnData
+    from mudata import MuData
+    from spatialdata import SpatialData
+else:
```
I also find the `else` block highly confusing. Never seen it before. Are we sure we don't just want to move it above the `if TYPE_CHECKING`?

falexwolf commented on lamindb/core/types.py:
```diff
@@ -7,3 +11,23 @@
+else:
+    from anndata import AnnData
+
+    if is_package_installed("mudata"):
```
Have we always called all these at package import-time? At least at some point I made an effort that some things are only imported if they're needed.

Always checking for installed and then importing spatialdata seems like something that won't scale well if we want to do it with more packages in the future?

falexwolf commented on lamindb/models/artifact.py:
```diff
@@ -364,7 +365,7 @@ def get_relative_path_to_directory(
 def get_artifact_kwargs_from_data(
     *,
-    data: Path | UPath | str | pd.DataFrame | AnnData | MuData,
+    data: Path | UPath | str | pd.DataFrame | ScverseDataStructures,
```
That's really nice!

falexwolf commented on tests/core/test_load.py:
```diff
@@ -69,3 +150,50 @@ def test_no_loader(zip_file):
+
+def test_load_svg(svg_filepath):
+    artifact = ln.Artifact(svg_filepath, key=str(svg_filepath))
+    result = artifact.load()
```
Why doesn't this error? Because the loader is configured to display an image and that will only work in ipython?

Zethson replied:
Correct. Else, it just returns a Path which I think is somewhat weird or inconsistent but not the scope if this PR.

```python
def load_svg(path: UPathStr) -> None | UPathStr:
    """Display `.svg` in ipython, otherwise return path."""
    if is_run_from_ipython:
        from IPython.display import SVG, display

        display(SVG(filename=path))
        return None
    else:
        return path
```

falexwolf commented on tests/core/data_fixtures.py:
```diff
@@ -0,0 +1,56 @@
+
+
+@pytest.fixture(scope="module")
+def adata():
```
Because there is no clean-up needed for simple in-memory objects, we don't need a fixture.

I'd say we should simply use the `core.datasets()` module consistently rather than introducing all these ad hoc datasets.

The good thing then is that we can use the same datasets in our docs.

Zethson replied:
Do we really want to clutter https://docs.lamin.ai/lamindb.core.datasets with even more test datasets? This will only grow...

Zethson replied:
I'll spin this out into an issue where I think more about how we can organize this with our goals:

1. Reuse test datasets in tutorials
2. Reuse fixtures across the many test functions
3. Keep everything sane and readable.

I think this goes beyond the scope of this PR now although I did a tiny tiny step here.

falexwolf replied:
> Do we really want to clutter https://docs.lamin.ai/lamindb.core.datasets with even more test datasets? This will only grow...

Yes, I think we should organize all datasets for unit tests and core docs properly in `core.datasets`. The alternative is that they're implicitly scattered all over the place which is much worse. I'm confident we can eliminate many of the datasets that date back years and aren't used anywhere anymore.

Koncopd commented on lamindb/core/storage/_tiledbsoma.py:
```diff
@@ -24,10 +24,10 @@
 
     if objpath.is_dir():
-        adata = load_anndata_zarr(objpath)
+        adata = load_zarr(objpath)
```
Also there are no checks for this upstream as it is called on an input in `save_tiledbsoma_experiment`.

Zethson replied:
https://github.com/laminlabs/lamindb/pull/2532/commits/27e961ca88b1334183d74ab02e6c977dfb763e45

is this a compromise? Then we retain some DRY but can be explicit if needed. We just don't always know what kind of  `.zarr` we get so I can't always use the appropriate one from the get go. I think this however should be okay.

Koncopd commented on lamindb/core/storage/_zarr.py:
```diff
@@ -48,29 +52,62 @@ def identify_zarr_type(
-    adata = read_zarr(open_obj)
-    return adata
+
+    actual_type = identify_zarr_type(storepath)
```
what i don't like here is that this makes a mapper (open_obj) again in addition to open_obj above. Maybe it can just return it?

Zethson replied:
Ahh makes sense. I added an additional method that only create the `open_obj` and I am then reusing that.

Koncopd commented on lamindb/core/storage/_zarr.py:
```diff
@@ -9,25 +9,44 @@
 
-def identify_zarr_type(
+
+def create_zarr_open_obj(
```
One last thing, i don't think this is needed as a separate function, you can move the code block inside `load_zarr`.

Zethson commented:
Addressed all comment except for moving some fixtures to datasets and very targeted tests for nested Exceptions. The PR got big so I'd like to get it in now. I'll follow up!
